### PR TITLE
Add quantization flag to BaseModel

### PR DESCRIPTION
**Pull request by blazekadam**
Haha, you did not expect this to be so easy huh? :D

I will add tests as soon as we migrate to TF2.0.

### CODE REVIEW

File: emloop_tensorflow/model.py [hunk 1]
```diff
@@ -67,6 +68,15 @@ def __init__(self,  # pylint: disable=too-many-arguments
...
             in the output.
 
+        .. warning::
+            Quantization aware training is rather experimental. You may run into many different issues among which
+            using unsupported ops is the most common one. Apart from specifying the quantize flag,
+            you need to register input and output tensors for the quantized graph with `_register_quantize_input` and
+            `_register_quantize_output` methods called from `_create_model` method.
+            For now, only TF 1.14 or 1.15 is supported. Both QA training graph with
+            fake quantization nodes and fully quantized .tfile flat buffer will be saved. By the way, the inputs
+            should be in [0, 255] range, otherwise they will be scaled badly. In conclusion, do not try this at home.
```
Comment (petrbel) [inline emloop_tensorflow/model.py]: Fun ending. I'd omit the last sentence but whatever.

File: emloop_tensorflow/model.py [hunk 2]
```diff
@@ -82,10 +92,27 @@ def __init__(self,  # pylint: disable=too-many-arguments
         :param clip_gradient: limit the absolute value of the gradient; set to None for no clipping
         :param profile: if true, profile the speed of model inference and save profiles to the specified log_dir
         :param keep_profiles: if true, profile the speed of model inference and save profiles to the specified log_dir
+        :param quantize: perform quantization-aware training from tf.contrib.quantize
+        :param quantize_delay: delay quantization-aware training by the specified number of iterations
+        :param args: additional args forwarded to :py:meth:`_create_model`
         :param kwargs: additional kwargs forwarded to :py:meth:`_create_model`
         """
+        if quantize and n_gpus > 1:
+            raise ValueError('Uh oh... quantization with n_gpus>1 is not exactly a great idea.'
```
Comment (petrbel) [inline emloop_tensorflow/model.py]: - I personally disapprove the level of (in)formality in our open-source project
- The content seems more like a warning than an exception terminating the training

File: emloop_tensorflow/model.py [hunk 3]
```diff
@@ -82,10 +92,27 @@ def __init__(self,  # pylint: disable=too-many-arguments
...
+                             'Most likely, it will not work or your PC will explode. Nothing can be ruled out.')
+        if quantize and restore_from is not None:
+            raise ValueError('Restoring a fake-quantized graph and trying to quantize it again or restoring a '
```
Comment (petrbel) [inline emloop_tensorflow/model.py]: this is actually even less than informal

File: emloop_tensorflow/model.py [hunk 4]
```diff
@@ -284,8 +309,49 @@ def save(self, name_suffix: str = '') -> str:
...
+                eval_sess.run(target_var.assign(self.session.run(source_tensor)))
+
+            logging.info(f'Converting model with inputs {self._quantize_inputs} and outputs f{self._quantize_outputs}')
+            converter = tf.lite.TFLiteConverter.from_session(eval_sess, self._quantize_inputs, self._quantize_outputs)
+            converter.inference_type = tf.uint8
+            # TODO: this will work only for inputs scaled to [0, positive_num] ... the user was warned though ...
+            converter.quantized_input_stats = {input_array: (0.0, 1.0) for input_array in converter.get_input_arrays()}
+            quantized = converter.convert()
+            quantized_model_path = path.join(self._log_dir, 'quantized.tflite')
+            with open(quantized_model_path, "wb") as f:
```
Comment (petrbel) [inline emloop_tensorflow/model.py]:
```suggestion
            with open(quantized_model_path, 'wb') as file:
```

File: emloop_tensorflow/model.py [hunk 5]
```diff
@@ -284,8 +309,49 @@ def save(self, name_suffix: str = '') -> str:
...
+                f.write(quantized)
```
Comment (petrbel) [inline emloop_tensorflow/model.py]:
```suggestion
                file.write(quantized)
```

File: emloop_tensorflow/model.py [hunk 6]
```diff
@@ -284,8 +309,49 @@ def save(self, name_suffix: str = '') -> str:
...
+            logging.info(f'Saving the quantized model to {quantized_model_path}')
```
Comment (petrbel) [inline emloop_tensorflow/model.py]:
```suggestion
            logging.info(f'Saving the quantized model to `{quantized_model_path}`')
```

**Review by petrbel — APPROVED** [review]
I added just some formatting suggestions and level-of-formality comments. Not sure I'm able to confirm it really works, though. Hope to see tests soon :) Btw I suggest @FloopCZ to look into the code as well since he's been into this way more than me

**Comment by blazekadam** [comment]
Thank you @petrbel ! I was a bit euphoric at that moment. Let me straighten that up.

**Comment by petrbel** [comment]
Thanks! Anyway, @FloopCZ pls make sure to read it.

**Pull request merged**